mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelFactory.kt

```diff
@@ -4,26 +4,76 @@ import android.support.v4.app.Fragment
+     * The initial state for the ViewModel. Override this if the initial state requires information from arguments, or the ViewModel owner.
+     * This function will take precedence over any secondary constructors defined in the state class, [S].
+     *
+     * The return value of this function will be transformed with any [PersistState] values before being used in [create].
+     *
+     * @return the initial state. If `null`, the state class constructors will be used for initial state creation.
      */
-    fun create(activity: FragmentActivity, state: S): BaseMvRxViewModel<S>
+    fun initialState(viewModelContext: ViewModelContext): S? = null
+
 }
 
 /**
- * Implement this in the companion object of a MvRxViewModel if your ViewModel needs more dependencies than just initial state.
- * If you only need the [FragmentActivity] then you can use the [MvRxViewModelFactory].
- * If all you need is initial state, you don't need to implement this at all.
+ * Creation context for the ViewModel. Includes the ViewModel store owner (either an activity or fragment), and fragment arguments
+ * set via [MvRx.KEY_ARG].
+ *
+ * For activity scoped ViewModels see [ActivityViewModelContext].
+ * For fragment scoped ViewModels see [FragmentViewModelContext].
+ *
+ * Never store a reference to this context, or the activity/fragment.
+ *
  */
-interface MvRxFragmentViewModelFactory<S : MvRxState> {
+sealed class ViewModelContext {
+    /**
+     * The activity which is using the ViewModel.
+     */
+    abstract val activity: FragmentActivity
+    /**
+     * Fragment arguments set via [MvRx.KEY_ARG].
+     */
+    abstract val rawArgs: Any?
+
     /**
-     * This will be called when your ViewModel needs to created. This *needs* to be annotated with [JvmStatic].
-     * @param state: The initial state for your ViewModel. This will be populated from fragment args and persisted state.
+     * Convenience method to type [rawArgs].
      */
-    fun create(fragment: Fragment, state: S): BaseMvRxViewModel<S>
+    @Suppress("UNCHECKED_CAST")
+    fun <A> args(): A = rawArgs as A
 }
+
+/**
+ * The [ViewModelContext] for a ViewModel created with an activity scope (`val viewModel by activityViewModel<MyViewModel>`). Although a fragment
+ * reference is available when an activity scoped ViewModel is first created, during process restoration, activity scoped ViewModels will be created
+ * _without_ a fragment reference, so it is only safe to reference the activity.
+ */
+class ActivityViewModelContext(
+    override val activity: FragmentActivity,
+    override val rawArgs: Any?
+) : ViewModelContext()
+
+/**
+ * The [ViewModelContext] for a ViewModel created with a
+ * fragment scope (`val viewModel by fragmentViewModel<MyViewModel>`).
+ */
+class FragmentViewModelContext(
+    override val activity: FragmentActivity,
+    override val rawArgs: Any?,
+    val fragment: Fragment
```

Comment:
@BenSchwab You could make a `<T : Fragment> fragment()` and `<A : Activity> activity()` function that could help with casting to your own app's Activity/Fragment